### PR TITLE
Typescript conversion, lint, tests, docs, examples, GitHub Actions, etc

### PR DESCRIPTION
**Pull request by huntharo**
- TypeScript rewrite
- Document all exported options and functions with jsdoc
- Add `eslint` and `prettier`
- Add `jest` and tests with `> 80%` statements, functions, and lines coverage
- Add GitHub Actions for continuous integration builds
- Validate parameters and log errors (but continue) when parameters are incompatible
- Store all state attached to one Express `app` instance to fix issues with in-process re-init of Express (e.g. during tests)
- `init` replaced with `trackConnections`
  - `trackConnections` requires the Express `app` to be passed as well as the `server` so that tracking can be isolated to a single instance instead of via global variables
  - Fixes `O(n^2)` performance in `trackConnections` due to storing connection references in an `Array`
  - Calls to `init` will log an error and do nothing as this method incurred performance penalty but was only needed when `force` was set to `true` (in other words `init` was likely being called by many who did not need to call it) - This only changes behavior in the case where the hard exit timer expires and will not change the behavior in any other case (other than improving performance)
  - `gracefulExitHandler` will log an error if `force`/`destroySocketsOnHardExit` is `true` but `trackConnections` was not called OR if `force`/`destroySocketsOnHardExit` is `false` but `trackConnections` was called
- `force` deprecated and replaced with `destroySocketsOnHardExit`. Settings passed to `force` will be copied to `destroySocketsOnHardExit`
- `suicideTimeout` deprecated and replaced with `hardExitTimeout`. Settings passed to `suicideTimeout` will be copied to `hardExitTimeout`

### CODE REVIEW

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    environment:
+      name: publish
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Use Node.js 16
+        uses: actions/setup-node@v2
+        with:
+          node-version: 16
```
Comment (ivolucien) [inline .github/workflows/publish.yml:20]: It seems like this and the .nvmrc should match, if only for consistency...

File: README.md [hunk 2]
```diff
@@ -1,99 +1,446 @@
 # express-graceful-exit
 
-Gracefully decline new requests while shutting down your application. A component that helps support zero downtime deploys for Node.js with [Express](http://expressjs.com/).
+Gracefully run last [Express](http://expressjs.com/) requests and respond with `Connection: close` response headers while optionally still allowing new connections for a time (e.g. for readiness probes) while shutting down your application.
```
Comment (ivolucien) [inline README.md:3]:
```suggestion
tl;dr - Gracefully handle final requests while your Express app shuts down.

Schedule app shutdown using a flexible set of options, while continuing to run last [Express](http://expressjs.com/) requests. Those last responses include `Connection: close` headers while optionally allowing new connections for a time (e.g. for readiness probes) while shutting down the application.
```

File: README.md [hunk 3]
```diff
@@ -1,99 +1,446 @@
...
+- [Zero Downtime Deploys](#zero-downtime-deploys)
+- [Compatibility](#compatibility)
+  - [v1.0.0 is backwards compatible with v0.x.x, with these notes:](#v100-is-backwards-compatible-with-v0xx-with-these-notes)
+  - [v0.X.X versions are backwards API compatible, with these minor behavior changes:](#v0xx-versions-are-backwards-api-compatible-with-these-minor-behavior-changes)
+- [Contributors](#contributors)
 
-```` bash
+# Why Graceful Shutdown
+
+Graceful shutdown is necessary if your infrastructure needs your application to exit itself to avoid terminating in-flight requests and generating errors / alarms.
```
Comment (ivolucien) [inline README.md:38]:
```suggestion
Graceful shutdown helps your application manage its own exit process, to complete in-flight requests and avoid error responses, dropped requests and/or error alarms.
```
Rewriting for concision and folks less fluent in English.

File: README.md [hunk 4]
```diff
@@ -1,99 +1,446 @@
...
+  - Waiting for the maximum response time (e.g. `server.timeout`) of one request will essentially guarantee that an instance has no further incoming connections
+  - This is not foolproof though as the choice of whether to use non-keep-alive connections is left to the client - Express would have to be configured with middleware to send a `connection: close` response header to every request, even those using `HTTP/1.1` or specifically sending a `connection: keep-alive` request header
+- Infrastructures that just wait a really long time for applications and clients that use HTTP/1.1 keep-alive connections
+  - An instance targeted for stopping is removed from the healthy instance pool, say, 30 minutes before the instance is to be stopped
+  - Simply waiting 30 minutes may result in that instance having no further incoming connections, but there is no guarantee of that
+  - It is possible for sockets to remain open indefintely so long as they continue to receive and respond to requests before either `server.timeout` or 1server.keepAliveTimeout` have expired
+  - This method may seem like it works, but it is never guaranteed to work
```
Comment (ivolucien) [inline README.md:71]: These sections are fabulous, but really need to be tl;dr'd, moved after the main docs, and/or moved to a linked page or file dedicated to this topic.

Most people would be intimidated by this mass of data that isn't yet telling them how to use the module.

File: README.md [hunk 5]
```diff
@@ -1,99 +1,446 @@
...
+
+# Zero Downtime Deploys
+
+This module handles the difficult work of gracefully closing sockets and the Express server in response to a developer choosing to initiate that process. It is possible and simple to install signal handlers, such as `SIGTERM` for Docker's default `STOPSIGNAL` without any additional dependencies and [src/examples/index.ts](src/examples/index.ts) shows how to accomplish this. Some prefer using another module to register signal handlers; this is fine, but care must be taken to ensure that the additional module does not add it's own shutdown timeouts that conflict with the timeouts set in `express-graceful-exit`.
 
-## Zero Downtime Deploys
+# Compatibility
 
-This module does not give you zero downtime deploys on its own. It enables the http server to exit gracefully, which when used with a module like naught can provide zero downtime deploys.
+## v1.0.0 is backwards compatible with v0.x.x, with these notes:
```
Comment (ivolucien) [inline README.md:429]:
```suggestion
## v1.0.0 is mostly backwards compatible with v0.x.x, with these breaking changes:
```

File: README.md [hunk 6]
```diff
@@ -1,99 +1,446 @@
...
+
+1. v1.0.0 Overall backwards compatible with the exception of uses cases that were passing `force: true` and calling `init` - In that one case `init` must be changed to `trackConnections` and both `server` and `app` must be passed to it
+1. v1.0.0 renames `init` to `trackConnections` and requires passing both the Express `server` and `app` - Note that `trackConnections` should ONLY be called if `force`/`destroySocketsOnHardExit` is `true` as the list of Express sockets is not necessary in any other case - As the usage of `init` is rarely necessary, calling `init` will log an error advising to call `trackConnections` instead and continue
+1. v1.0.0 renames `suicideTimeout` to `hardExitTimeout`. Values passed via `suicideTimeout` will be copied over top of `hardExitTimeout`
+1. v1.0.0 renames `force` to `destroySocketsOnHardExit`. Values passed via `force` will be copied over top of `destroySocketsOnHardExit`
```
Comment (ivolucien) [inline README.md:434]:
```suggestion
1. Apps that pass in `force: true` and call `init()` must change `init` to `trackConnections` and pass in both `server` and `app`
1. `init()` renamed to `trackConnections()` and requires passing both the Express `server` and `app` - Note that `trackConnections` should ONLY be called if `force`/`destroySocketsOnHardExit` is `true` as the list of Express sockets is not necessary in any other case - Calling `init` will log an error advising to call `trackConnections` instead and continue without error.
1. `suicideTimeout` option deprecated, replaced by `hardExitTimeout`. Either option works for now, but a future version (2.0?) will drop the `suicideTimeout` property.
1. `force` option deprecated, replaced by `destroySocketsOnHardExit`. Either option works for now, but a future version (2.0?) will drop the `force` property.
```

File: README.md [hunk 7]
```diff
@@ -1,99 +1,446 @@
...
 
-#### Author: [Jon Keating](http://twitter.com/emostar)
-This module was originally developed for Frafty (formerly www.frafty.com), a Daily Fantasy Sports site.
-#### Maintainer: [Ivo Havener](https://github.com/ivolucien)
+# Contributors
 
+- Original Author: [Jon Keating](http://twitter.com/emostar)
+  - This module was originally developed for Frafty (formerly www.frafty.com), a Daily Fantasy Sports site.
+- Maintainer: [Ivo Havener](https://github.com/ivolucien)
+- Contributor: [Harold Hunt](https://github.com/huntharo)
```
Comment (ivolucien) [inline README.md:446]: :+1:

File: src/lib/graceful-exit.ts [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+import { Configuration } from './configuration';
+import { Middleware } from './middleware';
+
+/**
+ * @deprecated 2022-02-25 - Renamed to `trackConnections`
+ *
+ * Tracking connections is not generally necessary and is only used to explicitly all incoming sockets to
+ * Express, which is not strictly necessary to do before exit.  As this functionality was not necessary
+ * for graceful shutdown handling this deprecated function has been left as a no-op to prevent applications
+ * that accidentally called it from having problems upon upgrade.
```
Comment (ivolucien) [inline src/lib/graceful-exit.ts:13]: Please reword this in a less opinionated way, along the lines of:

`init() is replaced by `trackConnections()` Unless you have specific need to track incoming sockets, calling `trackConnnections()` is unnecessary." 

And I suggest logging an Error object in `init()` to emphasize that it no longer works as expected.

**Review by ivolucien — CHANGES_REQUESTED** [review]
Some of these can be ignored if we don't merge this into the main repo.

**Comment by ivolucien** [comment]
Thanks to @huntharo for the submission, I'm sad to report that this PR isn't ready to be merged, has subtle breaking changes and is too opinionated to be incorporated into this widely used module.

I would be happy to merge a typescript rewrite that's aimed at a general audience with no breaking changes.
